components/brave_search/renderer/brave_search_render_frame_observer.cc

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_search/renderer/brave_search_render_frame_observer.h"
+
+#include <string>
+#include <vector>
+
+#include "base/no_destructor.h"
+#include "brave/components/brave_search/common/brave_search_utils.h"
+#include "brave/components/brave_search/renderer/brave_search_default_js_handler.h"
+#include "content/public/renderer/render_frame.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "url/gurl.h"
+
+namespace brave_search {
+
+BraveSearchRenderFrameObserver::BraveSearchRenderFrameObserver(
+    content::RenderFrame* render_frame,
+    int32_t world_id)
+    : RenderFrameObserver(render_frame), world_id_(world_id) {
+  native_javascript_handle_.reset(
+      new BraveSearchDefaultJSHandler(render_frame));
```

Comment:
we should create that object in `DidCreateScriptContext` as `render_frame` could be different or invalid while there. And if we bind remote to it, it will fail on mojo calls. Check a fix that is done here https://github.com/brave/brave-core/pull/9127/files#diff-bd0508db5780666e635128717e778f2a51486f055feb9f236d03aefdb1dc122cR35